In [32]:
import pandas as pd

# set_options to prevent DataFrame output from being truncated
pd.set_option('max_columns', 180)
pd.set_option('max_rows', 200000)
pd.set_option('max_colwidth', 5000)



In [3]:
game_log = pd.read_csv('game_log.csv')

game_log.shape




/dataquest/system/env/python3/lib/python3.4/site-packages/IPython/core/interactiveshell.py:2723: DtypeWarning: Columns (12,13,14,15,19,20,81,82,83,84,85,86,87,88,93,94,95,96,97,98,99,100,105,106,108,109,111,112,114,115,117,118,120,121,123,124,126,127,129,130,132,133,135,136,138,139,141,142,144,145,147,148,150,151,153,154,156,157,160) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


(171907, 161)

In [3]:
game_log.head()

date  number_of_game day_of_week v_name v_league  v_game_number h_name  \
0  18710504               0         Thu    CL1      NaN              1    FW1   
1  18710505               0         Fri    BS1      NaN              1    WS3   
2  18710506               0         Sat    CL1      NaN              2    RC1   
3  18710508               0         Mon    CL1      NaN              3    CH1   
4  18710509               0         Tue    BS1      NaN              2    TRO   

  h_league  h_game_number  v_score  h_score  length_outs day_night completion  \
0      NaN              1        0        2         54.0         D        NaN   
1      NaN              1       20       18         54.0         D        NaN   
2      NaN              1       12        4         54.0         D        NaN   
3      NaN              1       12       14         54.0         D        NaN   
4      NaN              1        9        5         54.0         D        NaN   

  forefeit protest park_id  attendance  length_minutes v_line_score  \
0      NaN     NaN   FOR01       200.0           120.0    000000000   
1      NaN     NaN   WAS01      5000.0           145.0    107000435   
2      NaN     NaN   RCK01      1000.0           140.0    610020003   
3      NaN     NaN   CHI01      5000.0           150.0    101403111   
4      NaN     NaN   TRO01      3250.0           145.0    000002232   

  h_line_score  v_at_bats  v_hits  v_doubles  v_triples  v_homeruns  v_rbi  \
0    010010000       30.0     4.0        1.0        0.0         0.0    0.0   
1    640113030       41.0    13.0        1.0        2.0         0.0   13.0   
2    010020100       49.0    11.0        1.0        1.0         0.0    8.0   
3    077000000       46.0    15.0        2.0        1.0         2.0   10.0   
4    101003000       46.0    17.0        4.0        1.0         0.0    6.0   

   v_sacrifice_hits  v_sacrifice_flies  v_hit_by_pitch  v_walks  \
0               0.0                0.0             0.0      1.0   
1               0.0                0.0             0.0     18.0   
2               0.0                0.0             0.0      0.0   
3               0.0                0.0             0.0      0.0   
4               0.0                0.0             0.0      2.0   

   v_intentional_walks  v_strikeouts  v_stolen_bases  v_caught_stealing  \
0                  NaN           6.0             1.0                NaN   
1                  NaN           5.0             3.0                NaN   
2                  NaN           1.0             0.0                NaN   
3                  NaN           1.0             0.0                NaN   
4                  NaN           0.0             1.0                NaN   

   v_grounded_into_double  v_first_catcher_interference  v_left_on_base  \
0                    -1.0                           NaN             4.0   
1                    -1.0                           NaN            12.0   
2                    -1.0                           NaN            10.0   
3                    -1.0                           NaN             7.0   
4                    -1.0                           NaN            12.0   

   v_pitchers_used  v_individual_earned_runs  v_team_earned_runs  \
0              1.0                       1.0                 1.0   
1              1.0                       6.0                 6.0   
2              1.0                       0.0                 0.0   
3              1.0                       6.0                 6.0   
4              1.0                       2.0                 2.0   

   v_wild_pitches  v_balks  v_putouts  v_assists  v_errors  v_passed_balls  \
0             0.0      0.0       27.0        9.0       0.0             3.0   
1             1.0      0.0       27.0       13.0      10.0             1.0   
2             2.0      0.0       27.0       12.0       8.0             5.0   
3             0.0      0.0       27.0       15.0      11.0             6.0   
4             0.0      0.0       

In [4]:
game_log.tail()

date  number_of_game day_of_week v_name v_league  v_game_number  \
171902  20161002               0         Sun    MIL       NL            162   
171903  20161002               0         Sun    NYN       NL            162   
171904  20161002               0         Sun    LAN       NL            162   
171905  20161002               0         Sun    PIT       NL            162   
171906  20161002               0         Sun    MIA       NL            161   

       h_name h_league  h_game_number  v_score  h_score  length_outs  \
171902    COL       NL            162        6        4         60.0   
171903    PHI       NL            162        2        5         51.0   
171904    SFN       NL            162        1        7         51.0   
171905    SLN       NL            162        4       10         51.0   
171906    WAS       NL            162        7       10         51.0   

       day_night completion forefeit protest park_id  attendance  \
171902         D        NaN      NaN     NaN   DEN02     27762.0   
171903         D        NaN      NaN     NaN   PHI13     36935.0   
171904         D        NaN      NaN     NaN   SFO03     41445.0   
171905         D        NaN      NaN     NaN   STL10     44615.0   
171906         D        NaN      NaN     NaN   WAS11     28730.0   

        length_minutes v_line_score h_line_score  v_at_bats  v_hits  \
171902           203.0   0200000202   1100100010       39.0    10.0   
171903           159.0    000001100    00100031x       33.0     8.0   
171904           184.0    000100000    23000002x       30.0     4.0   
171905           192.0    000020200    00100360x       35.0     9.0   
171906           216.0    000230020    03023002x       38.0    14.0   

        v_doubles  v_triples  v_homeruns  v_rbi  v_sacrifice_hits  \
171902        4.0        1.0         2.0    6.0               0.0   
171903        3.0        0.0         0.0    2.0               0.0   
171904        0.0        0.0         0.0    1.0               0.0   
171905        0.0        0.0         1.0    4.0               0.0   
171906        1.0        1.0         2.0    7.0               1.0   

        v_sacrifice_flies  v_hit_by_pitch  v_walks  v_intentional_walks  \
171902                0.0             1.0      4.0                  0.0   
171903                0.0             0.0      2.0                  0.0   
171904                0.0             0.0      2.0                  0.0   
171905                0.0             0.0      4.0                  0.0   
171906                0.0             0.0      3.0                  2.0   

        v_strikeouts  v_stolen_bases  v_caught_stealing  \
171902          12.0             2.0                1.0   
171903           9.0             1.0                1.0   
171904           7.0             0.0                0.0   
171905          11.0             0.0                1.0   
171906          10.0             1.0                1.0   

        v_grounded_into_double  v_first_catcher_interference  v_left_on_base  \
171902                     0.0                           0.0             8.0   
171903                     1.0                           0.0             6.0   
171904                     1.0                           0.0             4.0   
171905                     0.0                           0.0             8.0   
171906                     1.0                           0.0             8.0   

        v_pitchers_used  v_individual_earned_runs  v_team_earned_runs  \
171902              7.0                       4.0                 4.0   
171903              6.0                       3.0                 3.0   
171904              7.0                       7.0                 7.0   
171905              6.0                       8.0                 8.0   
171906              7.0                      10.0                10.0   

        v_wild_pitches  v_balks  v_putouts  v_assists  v_errors  \
171902             1.0      0.0       30.0       12.0       

In [84]:
!cat game_log_fields.txt

Field(s)  Meaning
    1     Date in the form "yyyymmdd"
    2     Number of game:
             "0" -- a single game
             "1" -- the first game of a double (or triple) header
                    including seperate admission doubleheaders
             "2" -- the second game of a double (or triple) header
                    including seperate admission doubleheaders
             "3" -- the third game of a triple-header
             "A" -- the first game of a double-header involving 3 teams
             "B" -- the second game of a double-header involving 3 teams
    3     Day of week  ("Sun","Mon","Tue","Wed","Thu","Fri","Sat")
  4-5     Visiting team and league
    6     Visiting team game number
          For this and the home team game number, ties are counted as
          games and suspended games are counted from the starting
          rather than the ending date.
  7-8     Home team and league
    9     Home team game number
10-11     Visiting and home team score (unquoted)


In [5]:
park_codes = pd.read_csv('park_codes.csv')
park_codes.shape

(252, 9)

In [6]:
park_codes.head()

park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                        aka       city state       start  \
0                                       NaN     Albany    NY  09/11/1880   
1                                       NaN    Altoona    PA  04/30/1884   
2             Edison Field; Anaheim Stadium    Anaheim    CA  04/19/1966   
3                                       NaN  Arlington    TX  04/21/1972   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX  04/11/1994   

          end league                                           notes  
0  05/30/1882     NL  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1  05/31/1884     UA                                             NaN  
2         NaN     AL                                             NaN  
3  10/03/1993     AL                                             NaN  
4         NaN     AL                                             NaN

In [7]:
person_codes = pd.read_csv("person_codes.csv")
person_codes.shape

(20494, 7)

In [88]:
person_codes.head()

id     last   first player_debut mgr_debut coach_debut ump_debut
0  aardd001  Aardsma   David   04/06/2004       NaN         NaN       NaN
1  aaroh101    Aaron    Hank   04/13/1954       NaN         NaN       NaN
2  aarot101    Aaron  Tommie   04/10/1962       NaN  04/06/1979       NaN
3  aased001     Aase     Don   07/26/1977       NaN         NaN       NaN
4  abada001     Abad    Andy   09/10/2001       NaN         NaN       NaN

In [8]:
team_codes = pd.read_csv("team_codes.csv")
team_codes.shape

(150, 8)

In [90]:
team_codes.head()

team_id league  start   end       city         nickname franch_id  seq
0     ALT     UA   1884  1884    Altoona  Mountain Cities       ALT    1
1     ARI     NL   1998     0    Arizona     Diamondbacks       ARI    1
2     BFN     NL   1879  1885    Buffalo           Bisons       BFN    1
3     BFP     PL   1890  1890    Buffalo           Bisons       BFP    1
4     BL1    NaN   1872  1874  Baltimore         Canaries       BL1    1

Brief summary about the files:
- game_log.csv
    - File has over 171 rows and 121 columns
    - Each row contains all game information including, date, place, teams, players, coaches, empires, final game score, and score per player
- park_codes.csv
    - Helper files for game_log.csv
    - Contains 252 rows and 9 columns with park information
    - Contains park_id also available on game_log.csv
- person_codes.csv
    - Helper file for game_log.csv
    - Contains 20,494 rows and 7 columns with player, coach, and empire information. 
    - The column id is used for each person on game_log.csv. Example: manager_id, winning_pitcher_id, and so forth
- team_codes.csv
    - Helper file for game_log.csv
    - Contains 150 rows and 8 columns with info about each team 
    - The team_id is used in game_log.csv

Baseball defensive position number
- There are nine field positions in baseball. Each one has a number. 
Example: 
    1. Pitcher
    2. Catcher
    3. First baseman
    4. Second baseman
    5. Third baseman
    6. Shortstop
    7. left fielder
    8. center fielder
    9. right fielder



In [33]:
import sqlite3

def run_query(q):
    with sqlite3.connect('mlb.db') as conn:
        return pd.read_sql(q, conn)
    
def execute_query(q):
    with sqlite3.connect('mlb.db') as conn:
        conn.execute('PRAGMA foreign_keys = ON;')
        conn.isolation_level = None
        conn.execute(q)



In [34]:
#Import files to SQLite
import sqlite3
con = sqlite3.connect('mlb.db')



            

In [ ]:
game_log.to_sql('game_log', con, index = False)

In [ ]:
park_codes.to_sql('park_codes', con, index = False)


In [ ]:
person_codes.to_sql('person_codes', con, index = False)


In [ ]:
team_codes.to_sql('team_codes', con, index = False)

In [5]:
q = """SELECT h_name,date,number_of_game,
      h_name || date ||number_of_game, game_id
  FROM game_log limit 5"""
run_query(q)

h_name      date  number_of_game h_name || date ||number_of_game  \
0    FW1  18710504               0                    FW1187105040   
1    WS3  18710505               0                    WS3187105050   
2    RC1  18710506               0                    RC1187105060   
3    CH1  18710508               0                    CH1187105080   
4    TRO  18710509               0                    TRO187105090   

        game_id  
0  FW1187105040  
1  WS3187105050  
2  RC1187105060  
3  CH1187105080  
4  TRO187105090

In [ ]:
q = "ALTER TABLE game_log ADD column game_id text"
execute_query(q)


In [ ]:
q = "UPDATE game_log SET game_id = h_name || date ||number_of_game"
execute_query(q)


![Alt text](/⁨Users⁩/⁨paulodossantos⁩/⁨Downloads⁩/dbschema.png)

In [ ]:
q = """
CREATE TABLE person
(person_id text primary key,
first_name text,
last_name text)
"""
execute_query(q)

In [12]:
q = "SELECT * FROM person_codes limit 5"
run_query(q)

id     last   first player_debut mgr_debut coach_debut ump_debut
0  aardd001  Aardsma   David   04/06/2004      None        None      None
1  aaroh101    Aaron    Hank   04/13/1954      None        None      None
2  aarot101    Aaron  Tommie   04/10/1962      None  04/06/1979      None
3  aased001     Aase     Don   07/26/1977      None        None      None
4  abada001     Abad    Andy   09/10/2001      None        None      None

In [ ]:
q = """
    INSERT INTO person(person_id, first_name, last_name)
        SELECT id, first, last from person_codes
    """
execute_query(q)

q = "SELECT * FROM person LIMIT 5"
run_query(q)

In [ ]:
q = """
CREATE TABLE park
(park_id text primary key,
name text,
nickname text,
city text,
state text,
notes text)
"""
execute_query(q)

In [13]:
q = "SELECT * FROM park_codes limit 5"
run_query(q)

park_id                           name  \
0   ALB01                 Riverside Park   
1   ALT01                  Columbia Park   
2   ANA01       Angel Stadium of Anaheim   
3   ARL01              Arlington Stadium   
4   ARL02  Rangers Ballpark in Arlington   

                                        aka       city state       start  \
0                                      None     Albany    NY  09/11/1880   
1                                      None    Altoona    PA  04/30/1884   
2             Edison Field; Anaheim Stadium    Anaheim    CA  04/19/1966   
3                                      None  Arlington    TX  04/21/1972   
4  The Ballpark in Arlington; Ameriquest Fl  Arlington    TX  04/11/1994   

          end league                                           notes  
0  05/30/1882     NL  TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1  05/31/1884     UA                                            None  
2        None     AL                                            None  
3  10/03/1993     AL                                            None  
4        None     AL                                            None

In [ ]:
q = """
    INSERT INTO park(park_id, name, nickname, city, state, notes)
        SELECT park_id, name, aka, city, state, notes from park_codes
    """
execute_query(q)

q = "SELECT * FROM park LIMIT 5"
run_query(q)

In [ ]:
q = """
CREATE TABLE league
(league_id text primary key,
name text)
"""
execute_query(q)

q = """
INSERT INTO league(league_id, name)
VALUES ('None',''),
       ('NL', 'National League'),
       ('AA', 'Double-A'),
       ('UA', 'Union Association'),
       ('PL', 'Players League'),
       ('AL', 'American League'),
       ('FL', 'Florida State League')
       
"""
execute_query(q)

In [14]:
q = "SELECT * from league"
run_query(q)

league_id                  name
0      None                      
1        NL       National League
2        AA              Double-A
3        UA     Union Association
4        PL        Players League
5        AL       American League
6        FL  Florida State League

In [96]:
q = """
CREATE TABLE appearance_type
(appearance_type_id text primary key,
name text,
category text)
"""
execute_query(q)



In [93]:
appearance_type = pd.read_csv('appearance_type.csv')
appearance_type.head()



appearance_type_id      name category
0                 O1  Batter 1  offense
1                 O2  Batter 2  offense
2                 O3  Batter 3  offense
3                 O4  Batter 4  offense
4                 O5  Batter 5  offense

In [94]:
appearance_type.to_sql('aat',con, index = False)

In [15]:
q = "SELECT * from aat limit 3"
run_query(q)

DatabaseError: Execution failed on sql 'SELECT * from aat limit 3': no such table: aat

In [98]:
q = """
INSERT INTO appearance_type(appearance_type_id, name, category)
    SELECT
        appearance_type_id, name, category
    FROM
        aat
       
"""
execute_query(q)

q = """
DROP TABLE aat
"""
execute_query(q)


q = "SELECT * from appearance_type limit 3"
run_query(q)

appearance_type_id      name category
0                 O1  Batter 1  offense
1                 O2  Batter 2  offense
2                 O3  Batter 3  offense

In [17]:
q = """
CREATE TABLE team
(team_id text primary key,
league_id text,
city text,
nickname text,
franch_id text)
"""
execute_query(q)

In [23]:
team_codes.to_sql('tc', con, index = False)

In [24]:
q = "SELECT * FROM tc LIMIT 5"
run_query(q)

team_id league  start   end       city         nickname franch_id  seq
0     ALT     UA   1884  1884    Altoona  Mountain Cities       ALT    1
1     ARI     NL   1998     0    Arizona     Diamondbacks       ARI    1
2     BFN     NL   1879  1885    Buffalo           Bisons       BFN    1
3     BFP     PL   1890  1890    Buffalo           Bisons       BFP    1
4     BL1   None   1872  1874  Baltimore         Canaries       BL1    1

In [25]:
q = """
INSERT OR IGNORE INTO team(team_id, league_id, city, nickname, franch_id)
    SELECT
        team_id, league, city, nickname, franch_id
    FROM
        tc
       
"""
execute_query(q)

In [26]:
q = "SELECT * FROM team LIMIT 5"
run_query(q)

team_id league_id       city         nickname franch_id
0     ALT        UA    Altoona  Mountain Cities       ALT
1     ARI        NL    Arizona     Diamondbacks       ARI
2     BFN        NL    Buffalo           Bisons       BFN
3     BFP        PL    Buffalo           Bisons       BFP
4     BL1      None  Baltimore         Canaries       BL1

In [28]:
q = "DROP TABLE tc"
execute_query(q)

In [41]:
q = """
CREATE TABLE game
(game_id text primary key,
date text,
number_of_game text,
park_id text,
length_outs decimal(10,2),
day Boolean,
completion text,
forfeit text,
protest text,
attendance int,
length_minutes decimal(10.2),
additional_info text,
acquisition_info text
)
"""
execute_query(q)

In [31]:
q = "SELECT * FROM game_log LIMIT 3"
run_query(q)

date  number_of_game day_of_week v_name v_league  v_game_number h_name  \
0  18710504               0         Thu    CL1     None              1    FW1   
1  18710505               0         Fri    BS1     None              1    WS3   
2  18710506               0         Sat    CL1     None              2    RC1   

  h_league  h_game_number  v_score  h_score  length_outs day_night completion  \
0     None              1        0        2         54.0         D       None   
1     None              1       20       18         54.0         D       None   
2     None              1       12        4         54.0         D       None   

  forefeit protest park_id  attendance  length_minutes v_line_score  \
0     None    None   FOR01       200.0           120.0    000000000   
1     None    None   WAS01      5000.0           145.0    107000435   
2     None    None   RCK01      1000.0           140.0    610020003   

  h_line_score  v_at_bats  v_hits  v_doubles  v_triples  v_homeruns  v_rbi  \
0    010010000       30.0     4.0        1.0        0.0         0.0    0.0   
1    640113030       41.0    13.0        1.0        2.0         0.0   13.0   
2    010020100       49.0    11.0        1.0        1.0         0.0    8.0   

   v_sacrifice_hits  v_sacrifice_flies  v_hit_by_pitch  v_walks  \
0               0.0                0.0             0.0      1.0   
1               0.0                0.0             0.0     18.0   
2               0.0                0.0             0.0      0.0   

  v_intentional_walks  v_strikeouts  v_stolen_bases v_caught_stealing  \
0                None           6.0             1.0              None   
1                None           5.0             3.0              None   
2                None           1.0             0.0              None   

   v_grounded_into_double v_first_catcher_interference  v_left_on_base  \
0                    -1.0                         None             4.0   
1                    -1.0                         None            12.0   
2                    -1.0                         None            10.0   

   v_pitchers_used  v_individual_earned_runs  v_team_earned_runs  \
0              1.0                       1.0                 1.0   
1              1.0                       6.0                 6.0   
2              1.0                       0.0                 0.0   

   v_wild_pitches  v_balks  v_putouts  v_assists  v_errors  v_passed_balls  \
0             0.0      0.0       27.0        9.0       0.0             3.0   
1             1.0      0.0       27.0       13.0      10.0             1.0   
2             2.0      0.0       27.0       12.0       8.0             5.0   

   v_double_plays  v_triple_plays  h_at_bats  h_hits  h_doubles  h_triples  \
0             0.0             0.0       31.0     4.0        1.0        0.0   
1             2.0             0.0       49.0    14.0        2.0        0.0   
2             0.0             0.0       36.0     7.0        2.0        1.0   

   h_homeruns  h_rbi  h_sacrifice_hits  h_sacrifice_flies  h_hit_by_pitch  \
0         0.0    2.0               0.0                0.0             0.0   
1         0.0   11.0               0.0                0.0             0.0   
2         0.0    2.0               0.0                0.0             0.0   

   h_walks h_intentional_walks  h_strikeouts  h_stolen_bases  \
0      1.0                None           0.0             0.0   
1     10.0                None           2.0             1.0   
2      0.0                None           3.0             5.0   

  h_caught_stealing  h_grounded_into_double h_first_catcher_interference  \
0              None                    -1.0                         None   
1              None                    -1.0                         None   
2              None                    -1.0                         None   

   h_left_on_base  h_pitchers_used  h_individual_earned_runs  \
0             3.0              1.0                       0.0   
1 

In [42]:
q = """INSERT OR IGNORE INTO game(game_id, date, number_of_game, 
    park_id, length_outs, day, completion, forfeit, protest, 
    attendance, length_minutes, additional_info, acquisition_info)
    
    SELECT 
        game_id, date, number_of_game, 
        park_id, length_outs, day_night, completion, forefeit, 
        protest, attendance, length_minutes, additional_info, 
        acquisition_info
    FROM
        game_log

"""
execute_query(q)

In [ ]:
q = "SELECT * FROM game LIMIT 3"
run_query(q)

In [6]:
# Get SQL to recreate a table
q = """
    SELECT sql FROM sqlite_master
    WHERE name = "game_log"
      AND type = "table";
    """
run_query(q)

sql
0  CREATE TABLE "game_log" (\n"date" INTEGER,\n  "number_of_game" INTEGER,\n  "day_of_week" TEXT,\n  "v_name" TEXT,\n  "v_league" TEXT,\n  "v_game_number" INTEGER,\n  "h_name" TEXT,\n  "h_league" TEXT,\n  "h_game_number" INTEGER,\n  "v_score" INTEGER,\n  "h_score" INTEGER,\n  "length_outs" REAL,\n  "day_night" TEXT,\n  "completion" TEXT,\n  "forefeit" TEXT,\n  "protest" TEXT,\n  "park_id" TEXT,\n  "attendance" REAL,\n  "length_minutes" REAL,\n  "v_line_score" TEXT,\n  "h_line_score" TEXT,\n  "v_at_bats" REAL,\n  "v_hits" REAL,\n  "v_doubles" REAL,\n  "v_triples" REAL,\n  "v_homeruns" REAL,\n  "v_rbi" REAL,\n  "v_sacrifice_hits" REAL,\n  "v_sacrifice_flies" REAL,\n  "v_hit_by_pitch" REAL,\n  "v_walks" REAL,\n  "v_intentional_walks" REAL,\n  "v_strikeouts" REAL,\n  "v_stolen_bases" REAL,\n  "v_caught_stealing" REAL,\n  "v_grounded_into_double" REAL,\n  "v_first_catcher_interference" REAL,\n  "v_left_on_base" REAL,\n  "v_pitchers_used" REAL,\n  "v_individual_earned_runs" REAL,\n  "v_team_earned_runs" REAL,\n  "v_wild_pitches" REAL,\n  "v_balks" REAL,\n  "v_putouts" REAL,\n  "v_assists" REAL,\n  "v_errors" REAL,\n  "v_passed_balls" REAL,\n  "v_double_plays" REAL,\n  "v_triple_plays" REAL,\n  "h_at_bats" REAL,\n  "h_hits" REAL,\n  "h_doubles" REAL,\n  "h_triples" REAL,\n  "h_homeruns" REAL,\n  "h_rbi" REAL,\n  "h_sacrifice_hits" REAL,\n  "h_sacrifice_flies" REAL,\n  "h_hit_by_pitch" REAL,\n  "h_walks" REAL,\n  "h_intentional_walks" REAL,\n  "h_strikeouts" REAL,\n  "h_stolen_bases" REAL,\n  "h_caught_stealing" REAL,\n  "h_grounded_into_double" REAL,\n  "h_first_catcher_interference" REAL,\n  "h_left_on_base" REAL,\n  "h_pitchers_used" REAL,\n  "h_individual_earned_runs" REAL,\n  "h_team_earned_runs" REAL,\n  "h_wild_pitches" REAL,\n  "h_balks" REAL,\n  "h_putouts" REAL,\n  "h_assists" REAL,\n  "h_errors" REAL,\n  "h_passed_balls" REAL,\n  "h_double_plays" REAL,\n  "h_triple_plays" REAL,\n  "hp_umpire_id" TEXT,\n  "hp_umpire_name" TEXT,\n  "1b_umpire_id" TEXT,\n  "1b_umpire_name" TEXT,\n  "2b_umpire_id" TEXT,\n  "2b_umpire_name" TEXT,\n  "3b_umpire_id" TEXT,\n  "3b_umpire_name" TEXT,\n  "lf_umpire_id" TEXT,\n  "lf_umpire_name" TEXT,\n  "rf_umpire_id" TEXT,\n  "rf_umpire_name" TEXT,\n  "v_manager_id" TEXT,\n  "v_manager_name" TEXT,\n  "h_manager_id" TEXT,\n  "h_manager_name" TEXT,\n  "winning_pitcher_id" TEXT,\n  "winning_pitcher_name" TEXT,\n  "losing_pitcher_id" TEXT,\n  "losing_pitcher_name" TEXT,\n  "saving_pitcher_id" TEXT,\n  "saving_pitcher_name" TEXT,\n  "winning_rbi_batter_id" TEXT,\n  "winning_rbi_batter_id_name" TEXT,\n  "v_starting_pitcher_id" TEXT,\n  "v_starting_pitcher_name" TEXT,\n  "h_starting_pitcher_id" TEXT,\n  "h_starting_pitcher_name" TEXT,\n  "v_player_1_id" TEXT,\n  "v_player_1_name" TEXT,\n  "v_player_1_def_pos" REAL,\n  "v_player_2_id" TEXT,\n  "v_player_2_name" TEXT,\n  "v_player_2_def_pos" REAL,\n  "v_player_3_id" TEXT,\n  "v_player_3_name" TEXT,\n  "v_player_3_def_pos" REAL,\n  "v_player_4_id" TEXT,\n  "v_player_4_name" TEXT,\n  "v_player_4_def_pos" REAL,\n  "v_player_5_id" TEXT,\n  "v_player_5_name" TEXT,\n  "v_player_5_def_pos" REAL,\n  "v_player_6_id" TEXT,\n  "v_player_6_name" TEXT,\n  "v_player_6_def_pos" REAL,\n  "v_player_7_id" TEXT,\n  "v_player_7_name" TEXT,\n  "v_player_7_def_pos" REAL,\n  "v_player_8_id" TEXT,\n  "v_player_8_name" TEXT,\n  "v_player_8_def_pos" REAL,\n  "v_player_9_id" TEXT,\n  "v_player_9_name" TEXT,\n  "v_player_9_def_pos" REAL,\n  "h_player_1_id" TEXT,\n  "h_player_1_name" TEXT,\n  "h_player_1_def_pos" REAL,\n  "h_player_2_id" TEXT,\n  "h_player_2_name" TEXT,\n  "h_player_2_def_pos" REAL,\n  "h_player_3_id" TEXT,\n  "h_player_3_name" TEXT,\n  "h_player_3_def_pos" REAL,\n  "h_player_4_id" TEXT,\n  "h_player_4_name" TEXT,\n  "h_player_4_def_pos" REAL,\n  "h_player_5_id" TEXT,\n  "h_player_5_name" TEXT,\n  "h_player_5_def_pos" REAL,\n  "h_player_6_id" TEXT,\n  "h_player_6_name" TEXT,\n  "h_player_6_def_pos" REAL,\n  "h_player_7_id" TEXT,\n  "h_player_7_name" TEXT,\n  "h_playe

In [7]:
q = """
CREATE TABLE team_appearance(
team_id text,
game_id text,
home boolean,
league_id text,
score integer,
line_score text,
at_bats REAL,
hits REAL,
doubles REAL,
triples REAL,
homeruns REAL,
rbi REAL,
sacrifice_hits REAL,
sacrifice_flies REAL,
hit_by_pitch REAL,
walks REAL,
intentional_walks REAL,
strikeouts REAL,
stolen_bases REAL,
caught_stealing REAL,
grounded_into_double REAL,
first_catcher_interference REAL,
left_on_base REAL,
pitchers_used REAL,
individual_earned_runs REAL,
team_earned_runs REAL,
wild_pitches REAL,
balks REAL,
putouts REAL,
assists REAL,
errors REAL,
passed_balls REAL,
double_plays REAL,
triple_plays REAL
)

"""
execute_query(q)

In [13]:
q = """
INSERT OR IGNORE INTO team_appearance
    SELECT
        v_name,
        game_id,
        0 AS home,
        v_league, 
        v_score,
        v_line_score, 
        v_at_bats,
        v_hits, 
        v_doubles,
        v_triples,
        v_homeruns, 
        v_rbi, 
        v_sacrifice_hits,
        v_sacrifice_flies,
        v_hit_by_pitch,
        v_walks, 
        v_intentional_walks,
        v_strikeouts, 
        v_stolen_bases, 
        v_caught_stealing, 
        v_grounded_into_double, 
        v_first_catcher_interference, 
        v_left_on_base, 
        v_pitchers_used, 
        v_individual_earned_runs, 
        v_team_earned_runs, 
        v_wild_pitches, 
        v_balks, 
        v_putouts, 
        v_assists, 
        v_errors, 
        v_passed_balls,
        v_double_plays, 
        v_triple_plays
    FROM
        game_log
    
"""
execute_query(q)


In [10]:
q = """
INSERT OR IGNORE INTO team_appearance
    SELECT
        h_name,
        game_id,
        1 AS home,
        h_league, 
        h_score,
        h_line_score, 
        h_at_bats,
        h_hits, 
        h_doubles,
        h_triples,
        h_homeruns, 
        h_rbi, 
        h_sacrifice_hits,
        h_sacrifice_flies,
        h_hit_by_pitch,
        h_walks, 
        h_intentional_walks,
        h_strikeouts, 
        h_stolen_bases, 
        h_caught_stealing, 
        h_grounded_into_double, 
        h_first_catcher_interference, 
        h_left_on_base, 
        h_pitchers_used, 
        h_individual_earned_runs, 
        h_team_earned_runs, 
        h_wild_pitches, 
        h_balks, 
        h_putouts, 
        h_assists, 
        h_errors, 
        h_passed_balls,
        h_double_plays, 
        h_triple_plays
    FROM
        game_log
"""
execute_query(q)

In [14]:
q = "SELECT * FROM team_appearance WHERE Home = 0 LIMIT 10"
run_query(q)

team_id       game_id  home league_id  score    line_score  at_bats  hits  \
0     CL1  FW1187105040     0      None      0     000000000     30.0   4.0   
1     BS1  WS3187105050     0      None     20     107000435     41.0  13.0   
2     CL1  RC1187105060     0      None     12     610020003     49.0  11.0   
3     CL1  CH1187105080     0      None     12     101403111     46.0  15.0   
4     BS1  TRO187105090     0      None      9     000002232     46.0  17.0   
5     CH1  CL1187105110     0      None     18      12120534     41.0  15.0   
6     WS3  CL1187105130     0      None     12     141020004     42.0   9.0   
7     CH1  FW1187105130     0      None     14     053210012     45.0  17.0   
8     WS3  FW1187105150     0      None      6     030100101     42.0   8.0   
9     TRO  BS1187105160     0      None     29  302604(11)30     64.0  26.0   

   doubles  triples  homeruns   rbi  sacrifice_hits  sacrifice_flies  \
0      1.0      0.0       0.0   0.0             0.0              0.0   
1      1.0      2.0       0.0  13.0             0.0              0.0   
2      1.0      1.0       0.0   8.0             0.0              0.0   
3      2.0      1.0       2.0  10.0             0.0              0.0   
4      4.0      1.0       0.0   6.0             0.0              0.0   
5      1.0      3.0       3.0  10.0             0.0              0.0   
6      2.0      0.0       0.0   5.0             0.0              0.0   
7      5.0      2.0       0.0  10.0             0.0              0.0   
8      0.0      0.0       0.0   2.0             0.0              0.0   
9      3.0      1.0       0.0  26.0             0.0              0.0   

   hit_by_pitch  walks intentional_walks  strikeouts  stolen_bases  \
0           0.0    1.0              None         6.0           1.0   
1           0.0   18.0              None         5.0           3.0   
2           0.0    0.0              None         1.0           0.0   
3           0.0    0.0              None         1.0           0.0   
4           0.0    2.0              None         0.0           1.0   
5           0.0    8.0              None         1.0           0.0   
6           0.0    1.0              None         1.0           1.0   
7           0.0    1.0              None         1.0           2.0   
8           0.0    1.0              None         1.0           1.0   
9           0.0    2.0              None         0.0           3.0   

  caught_stealing  grounded_into_double first_catcher_interference  \
0            None                  -1.0                       None   
1            None                  -1.0                       None   
2            None                  -1.0                       None   
3            None                  -1.0                       None   
4            None                  -1.0                       None   
5            None                  -1.0                       None   
6            None                  -1.0                       None   
7            None                  -1.0                       None   
8            None                  -1.0                       None   
9            None                  -1.0                       None   

   left_on_base  pitchers_used  individual_earned_runs  team_earned_runs  \
0           4.0            1.0                     1.0               1.0   
1          12.0            1.0                     6.0               6.0   
2          10.0            1.0                     0.0               0.0   
3           7.0            1.0                     6.0               6.0   
4          12.0            1.0                     2.0               2.0   
5           7.0            1.0                     4.0               4.0   
6           4.0            1.0                     2.0               2.0   
7           5.0            1.0                     2.0               2.0   
8          10.0            1.0                     4.0               4.0   
9          10.0            1.0       

In [8]:
q = "delete from team_appearance"
#execute_query(q)

In [17]:
q = """
    CREATE TABLE person_appearance(
    appearance_id INTEGER PRIMARY KEY AUTOINCREMENT,
    person_id text,
    team_id text,
    game_id text,
    appearance_type_id TEXT
    )
"""
execute_query(q)

In [37]:
q = """INSERT OR IGNORE INTO person_appearance (
    game_id,
    team_id,
    person_id,
    appearance_type_id
)
SELECT
    game_id,
    NULL,
    lf_umpire_id,
    "ULF"
FROM game_log
WHERE lf_umpire_id IS NOT NULL
UNION

    SELECT
        game_id,
        NULL,
        rf_umpire_id,
        "URF"
    FROM game_log
    WHERE rf_umpire_id IS NOT NULL

UNION

    SELECT
        game_id,
        v_name,
        v_manager_id,
        "MM"
    FROM game_log
    WHERE v_manager_id IS NOT NULL

UNION

    SELECT
        game_id,
        h_name,
        h_manager_id,
        "MM"
    FROM game_log
    WHERE h_manager_id IS NOT NULL

UNION

    SELECT
        game_id,
        CASE
            WHEN h_score > v_score THEN h_name
            ELSE v_name
            END,
        winning_pitcher_id,
        "AWP"
    FROM game_log
    WHERE winning_pitcher_id IS NOT NULL


"""
execute_query(q)

In [40]:
template = """
INSERT OR IGNORE INTO person_appearance (
    game_id,
    team_id,
    person_id,
    appearance_type_id
) 
    SELECT
        game_id,
        {hv}_name,
        {hv}_player_{num}_id,
        "O{num}"
    FROM game_log
    WHERE {hv}_player_{num}_id IS NOT NULL

UNION

    SELECT
        game_id,
        {hv}_name,
        {hv}_player_{num}_id,
        "D" || CAST({hv}_player_{num}_def_pos AS INT)
    FROM game_log
    WHERE {hv}_player_{num}_id IS NOT NULL;
"""

#execute_query(c1)
#execute_query(c2)

for hv in ["h","v"]:
    for num in range(1,10):
        query_vars = {
            "hv": hv,
            "num": num
        }
        # run commmand is a helper function which runs
        # a query against our database.
        execute_query(template.format(**query_vars))


In [19]:
q = "SELECT * FROM appearance_type"
run_query(q)

appearance_type_id                name category
0                  O1            Batter 1  offense
1                  O2            Batter 2  offense
2                  O3            Batter 3  offense
3                  O4            Batter 4  offense
4                  O5            Batter 5  offense
5                  O6            Batter 6  offense
6                  O7            Batter 7  offense
7                  O8            Batter 8  offense
8                  O9            Batter 9  offense
9                  D1             Pitcher  defense
10                 D2             Catcher  defense
11                 D3            1st Base  defense
12                 D4            2nd Base  defense
13                 D5            3rd Base  defense
14                 D6           Shortstop  defense
15                 D7          Left Field  defense
16                 D8        Center Field  defense
17                 D9         Right Field  defense
18                D10    Unknown Position  defense
19                UHP          Home Plate   umpire
20                U1B          First Base   umpire
21                U2B         Second Base   umpire
22                U3B          Third Base   umpire
23                ULF          Left Field   umpire
24                URF         Right Field   umpire
25                 MM             Manager  manager
26                AWP     Winning Pitcher    award
27                ALP      Losing Pitcher    award
28                ASP      Saving Pitcher    award
29                AWB  Winning RBI Batter    award
30                PSP    Starting Pitcher  pitcher

In [41]:
q = """SELECT
    COUNT(1)
FROM person_appearance

"""
run_query(q)

COUNT(1)
0   5554417

In [43]:
q = """DROP TABLE game_log;"""
execute_query(q)

q = """DROP TABLE park_codes;"""
execute_query(q)


q = """DROP TAblE team_codes;"""
execute_query(q)


q = """DROP TABLE person_codes"""
execute_query(q)

